python/sdk/merlin/pyfunc.py

```diff
@@ -109,7 +117,190 @@ def upiv1_infer(self, request: upi_pb2.PredictValuesRequest,
+        is loaded.
+
+        :param context: A :class:`~PythonModelContext` instance containing artifacts that the model
+                        can use to perform inference
+        """
+        self.initialize(context.artifacts)
+
+    def predict(self, context, model_input):
+        """
+        Implementation of PythonModel `predict` method. This method evaluates model input and produces model output.
+
+        :param context: A :class:`~PythonModelContext` instance containing artifacts that the model
+                        can use to perform inference
+        :param model_input: A pyfunc-compatible input for the model to evaluate.
+        """
+        extra_args = model_input.get(PYFUNC_EXTRA_ARGS_KEY, {})
+        input = model_input.get(PYFUNC_MODEL_INPUT_KEY, {})
+        protocol = model_input.get(PYFUNC_PROTOCOL_KEY, Protocol.HTTP_JSON)
+
+        if protocol == Protocol.HTTP_JSON:
+            if extra_args is not None:
+                return self._do_http_predict(input, **extra_args)
+            return self._do_http_predict(input)
+        elif protocol == Protocol.UPI_V1:
+            grpc_context = model_input.get(PYFUNC_GRPC_CONTEXT, None)
+            ml_model_input = self.upiv1_preprocess(input, grpc_context)
+            ml_model_output = self.ml_predict(ml_model_input)
+            final_output = self.upiv1_postprocess(ml_model_output, input)
+            return PyFuncOutput(upi_response=final_output, model_input=ml_model_input, model_output=ml_model_output)
+        else:
+            raise NotImplementedError(f"protocol {protocol} is not supported")
 
+    def _do_http_predict(self, model_input, **kwargs):
+        ml_model_input = self.preprocess(model_input, **kwargs)
+        ml_model_output = self.ml_predict(ml_model_input)
+        final_output = self.postprocess(ml_model_output, model_input)
+
+        return PyFuncOutput(http_response=final_output, model_input=ml_model_input, model_output=ml_model_output)
+
+    @abstractmethod
+    def initialize(self, artifacts: dict):
+        """
+        Implementation of PyFuncModel can specify initialization step which
+        will be called one time during model initialization.
+
+        :param artifacts: dictionary of artifacts passed to log_model method
+        """
+        pass
+
+
+    @abstractmethod
+    def preprocess(self, request: dict, **kwargs) -> ModelInput:
+        """
+        preprocess is the method that doing preprocessing before calling the ml framework for prediction.
+        This method is called when user use HTTP_JSON protocol
+        :param request: raw input to the model (dict)
+        :return: model input, this model input must already have all the features that required for model prediction
+        """
+        raise NotImplementedError("preprocess is not implemented")
+
+    @abstractmethod
+    def ml_predict(self, model_input: ModelInput) -> ModelOutput:
```

Comment:
To match the `PyFuncV2Model` class, should we call this method `infer` ?

Btw, on a related note, I don't see the [preprocess](https://github.com/caraml-dev/merlin/blob/main/python/sdk/merlin/pyfunc.py#L166) and [postprocess](https://github.com/caraml-dev/merlin/blob/main/python/sdk/merlin/pyfunc.py#L177) methods of the `PyFuncV2Model` class being called anywhere. Am I missing something?

Reply:
honestly I don't have enough context why there is `preprocess` and postprocess` in the `PyFuncV2Model` in the comment itself stated that it is not called during batch inference. Actually the only method that is called is `infer` same like `PyFuncModel`

I'm prefer not to use `infer` method because user uses this to do all of the preprocessing, ml prediction, and postprocess in this method in `PyFuncModel`,  use `infer` method will add ambiguity IMO

Reply:
> honestly I don't have enough context why there is preprocess and postprocessin thePyFuncV2Modelin the comment itself stated that it is not called during batch inference. Actually the only method that is called isinfersame likePyFuncModel`

@ariefrahmansyah, any clue about this?

Reply:
> I'm prefer not to use infer method because user uses this to do all of the preprocessing, ml prediction, and postprocess in this method in PyFuncModel, use infer method will add ambiguity IMO

This is a valid concern. But I still think `infer` is better, for consistency. I also feel that `ml_predict` can be a bit restrictive (what if the model is non-ML based?). There is also no harm if the user does the pre and post process inside `infer`, unless they are enabling model monitoring, right? At the end of the day, if the user choses to use the V3 model version, I think the inline comments on each of the methods can clearly capture the purpose; for the `infer` method, we can mention that pre/post processing is recommended to be done in the other respective methods.

Reply:
-> Why there is a preprocess and postprocess in the PyFuncV2?

Because when we first design batch prediction case in Merlin, we envisioned PyFuncV2 to be able to be implemented in both online and batch case.

For example, case 1: User implements preprocess, infer, and postprocess methods
- This model implementation is reusable for both online prediction and batch prediction
- If it’s deployed as online prediction the request will be served using preprocess->infer->postprocess sequence.
- If it’s deployed as batch prediction the infer method will be used to calculate the inference result. 
This model implementation is intended for reusability, but sacrifices performance slightly for online prediction.

<img width="600" alt="image" src="https://github.com/caraml-dev/merlin/assets/8122852/94a1b0e0-ee18-43dd-a2a0-2308768638a0">

Case 2: User implements raw_infer, preprocess, infer, and postprocess methods
- This model implementation is also reusable for both online prediction and batch prediction
- Compared to case 1, the online prediction will only use the raw_infer method to serve incoming requests. This is intended so that users shouldn’t need to convert the incoming request into pandas Dataframe which might incur additional latency. 
- This model implementation is intended for both performance and reusability, but with extra duplication (since the user will have to implement model inference in both raw_infer and infer methods)
- 
<img width="600" alt="image" src="https://github.com/caraml-dev/merlin/assets/8122852/9acc8e6a-51ff-4c58-ad7a-7911d61c7350">

Reply:
However, preprocess, postprocess, and raw_infer are not called by the current pyfunc server. So to implement it, users need to call preprocess and postprocess from their infer method. And for raw_infer, there's no way for users to call it so it's actually not implemented by our users.

Reply:
Got it. Thanks for the detailed explanation, @ariefrahmansyah ! Given that the current state of the code for PyFuncV2 is only partially implementing these, do you see any issues removing these abstract methods (`preprocess`, `postprocess` and `raw_infer`) from PyFuncV2 entirely? That will prevent any confusion about their existence. If, by any chance, the user has implemented these in their concrete class, it should still continue to work.

Reply:
-> do you see any issues removing these abstract methods (preprocess, postprocess and raw_infer) from PyFuncV2 entirely?

No, I don't foresee any issue. It should be safe to remove them.

Reply:
Thanks for the confirmation, @ariefrahmansyah! @tiopramayudi, if you don't have any objections either, please go ahead and remove those methods from the PyFuncV2 class. Thanks!